Doc/regions.md

```diff
@@ -0,0 +1,567 @@
+Here we present the idea as a reverse polish evaluator.
+```python
+>>> x = 1
+>>> y = 2
+>>> calc("4 x y + *")
+```
+The idea is to interpolate the `x` and `y` into the string while evaluating it.
+```
+def calc(str):
+  ctxt = sys._getframe().f_back
+  ops = str.split()
+  stack = []
+  for op in ops:
+    try: 
+      stack.append(int(op))
+    except ValueError:
+      if op == "+":
+        stack.append(stack.pop() + stack.pop())
+      elif op == "*":
+        stack.append(stack.pop() * stack.pop())
+      else:    
+        stack.append(ctxt.f_locals[op])      
+  return stack.pop()
+```
+This is a simple example, but it shows how the frame object can be used to access the local variables in the calling scope.
+```
+>>> x = 1
+>>> y = 2
+>>> calc("4 x y + *")
+12
+```
+
+The numpy library is pretty popular, so this is a good example of how to use the frame object in a way that is not debugging.
+
+
+Note that the frame object is mostly read-only, but does contain an interesting operation to set the executing line number of a frame.
+```C
+static int
+frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno, void *Py_UNUSED(ignored))
+```
+This seems very bad to call from another thread/interpreter.
+
+Ultimately, a frame object should never escape the thread/interpreter that it was created in.
+
+
+
+## Regions in Veronapy.
+
+The previous section illustrates the many temporary objects that Python creates to handle variables.
+To support this, we need our region object model to support complex local structures for indexing into regions.
+To achieve this, we will use three kinds of object ownership:
+
+* Local object 
+* Immutable
+* Regions
+
+The "local objects" are used to represent the many local structures that Python creates such as function objects, frame objects and cells.
+These objects are typically not shared between threads, so normally should not require any region tracking.
+However, we need to support local objects becoming immutable to allow them to be shared, and suppport for groups of objects to be moved between interpreters.
+
+To pass mutable states between threads, we need to introduce a new construct to Python: "regions".
+A region is a group of objects that can move together across threads
+and can be accessed by at most one interpreter at a time.
+This removes the need for locks to protect the objects in the region.
+The region can be passed between threads, and the receiving interpreter can access the objects in the region.
+
+To efficiently support regions, we impose the following invariants that the runtime is responsible for maintaining:
+
+* Local objects can reach anything
+* Immutable objects can only reach other immutable objects.
+* Objects in a region can only reach other objects in the same region, immutable objects, or linear region entry points.
+* Regions track the number of incoming references from local objects.
```

Comment:
This concept of tracking incoming references seems to be the crux of your algorithms, and it makes me a little worried. During normal execution the interpreter frequently adds and removes temporary references (e.g. constructing or unpacking tuples).

Reply:
Thanks for the comments @gvanrossum ! Note that this check is only done at the point when we want to send a region. Before that point, we don’t want to restrict (at least so we believe) references any more than Python does today. 

The check is then performed by the “sending thread” which will therefore not be modifying reference counts at the same time. Concurrent modifications are not possible because there is no shared mutable state in this model. 

The check is necessary to ensure that when some object graph is sent off, no aliases to the graph remains with the sender. If that was possible, we could have data races on those objects, and would have to do atomic reference counting etc. like in PEP703.